usr/man/man/usr/share/man/man1/cktime.1
'\" te
.\"  Copyright 1989 AT&T  Copyright (c) 1992, Sun Microsystems, Inc.  All Rights Reserved
.TH cktime 1 "14 Sep 1992" "SunOS 5.11" "User Commands"
.SH NAME
cktime, errtime, helptime, valtime \- display a prompt; verify and return a time of day
.SH SYNOPSIS
.LP
.nf
\fBcktime\fR [\fB-Q\fR] [\fB-W\fR \fIwidth\fR] [\fB-f\fR \fIformat\fR] [\fB-d\fR \fIdefault\fR] [\fB-h\fR \fIhelp\fR] 
     [\fB-e\fR \fIerror\fR] [\fB-p\fR \fIprompt\fR] [\fB-k\fR \fIpid\fR [\fB-s\fR \fIsignal\fR]]
.fi

.LP
.nf
\fB/usr/sadm/bin/errtime\fR [\fB-W\fR \fIwidth\fR] [\fB-e\fR \fIerror\fR] [\fB-f\fR \fIformat\fR]
.fi

.LP
.nf
\fB/usr/sadm/bin/helptime\fR [\fB-W\fR \fIwidth\fR] [\fB-h\fR \fIhelp\fR] [\fB-f\fR \fIformat\fR]
.fi

.LP
.nf
\fB/usr/sadm/bin/valtime\fR [\fB-f\fR \fIformat\fR] \fIinput\fR
.fi

.SH DESCRIPTION
.sp
.LP
The \fBcktime\fR utility prompts a user and validates the response. It defines, among other things, a prompt message whose response should be a time, text for help and error messages, and a default value (which is returned if the user responds with a RETURN). The user response must match the defined format for the time of day.
.sp
.LP
All messages are limited in length to 70 characters and are formatted automatically. Any white space used in the definition (including \fBNEWLINE)\fR is stripped. The \fB-W\fR option cancels the automatic formatting. When a tilde is placed at the beginning or end of a message definition, the default text is inserted at that point, allowing both custom text and the default text to be displayed.
.sp
.LP
If the prompt, help or error message is not defined, the default message (as defined under \fBNOTES\fR) is displayed.
.sp
.LP
Three visual tool modules are linked to the \fBcktime\fR command. They are \fBerrtime\fR (which formats and displays an error message), \fBhelptime\fR (which formats and displays a help message), and \fBvaltime\fR (which validates a response). These modules should be used in conjunction with  \fBFML\fR objects. In this instance, the \fBFML\fR object defines the prompt. When \fBformat\fR is defined in the \fBerrtime\fR and \fBhelptime\fR modules, the messages will describe the expected format.
.SH OPTIONS
.sp
.LP
The following options are supported:
.sp
.ne 2
.mk
.na
\fB\fB-d\fR\fI default\fR\fR
.ad
.RS 14n
.rt  
Defines the default value as \fIdefault\fR. The default is not validated and so does not have to meet any criteria.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-e\fR \fIerror\fR\fR
.ad
.RS 14n
.rt  
Defines the error message as \fIerror\fR.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-f\fR \fIformat\fR\fR
.ad
.RS 14n
.rt  
Specifies the format against which the input will be verified. Possible formats and their definitions are: 
.sp
.in +2
.nf
\fB%H\fR  =  hour (00 - 23)
\fB%I\fR  =  hour (00 - 12)
\fB%M\fR  =  minute (00 - 59)
\fB%p\fR  =  ante meridian or post meridian
\fB%r\fR  =  time as \fB%I:%M:%S %p\fR
\fB%R\fR  =  time as \fB%H:%M\fR (the default format)
\fB%S\fR  =  seconds (00 - 59)
\fB%T\fR  =  time as \fB%H:%M:%S\fR
.fi
.in -2
.sp

.RE

.sp
.ne 2
.mk
.na
\fB\fB-h\fR \fIhelp\fR\fR
.ad
.RS 14n
.rt  
Defines the help messages as \fIhelp\fR.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-k\fR\fI pid\fR\fR
.ad
.RS 14n
.rt  
Specifies that process \fBID\fR \fIpid\fR is to be sent a signal if the user chooses to abort.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-p\fR\fI prompt\fR\fR
.ad
.RS 14n
.rt  
Defines the prompt message as \fIprompt\fR.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-Q\fR\fR
.ad
.RS 14n
.rt  
Specifies that quit will not be allowed as a valid response.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-s\fR \fIsignal\fR\fR
.ad
.RS 14n
.rt  
Specifies that the process \fBID\fR \fIpid\fR defined with the \fB-k\fR option is to be sent signal \fIsignal\fR when quit is chosen.  If no signal is  specified, \fBSIGTERM\fR is used.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-W\fR\fI width\fR\fR
.ad
.RS 14n
.rt  
Specifies that prompt, help and error messages will be formatted to a line length of \fIwidth\fR.
.RE

.SH OPERANDS
.sp
.LP
The following operand is supported:
.sp
.ne 2
.mk
.na
\fB\fIinput\fR\fR
.ad
.RS 9n
.rt  
Input to be verified against format criteria.
.RE

.SH EXIT STATUS
.sp
.LP
The following exit values are returned:
.sp
.ne 2
.mk
.na
\fB\fB0\fR\fR
.ad
.RS 5n
.rt  
Successful execution.
.RE

.sp
.ne 2
.mk
.na
\fB\fB1\fR\fR
.ad
.RS 5n
.rt  
\fBEOF\fR on input, or negative width on  \fB-W\fR option,  or usage error .
.RE

.sp
.ne 2
.mk
.na
\fB\fB3\fR\fR
.ad
.RS 5n
.rt  
User termination (quit) .
.RE

.sp
.ne 2
.mk
.na
\fB\fB4\fR\fR
.ad
.RS 5n
.rt  
Garbled format argument.
.RE

.SH ATTRIBUTES
.sp
.LP
See \fBattributes\fR(5) for descriptions of the following attributes:
.sp

.sp
.TS
tab() box;
cw(2.75i) |cw(2.75i) 
lw(2.75i) |lw(2.75i) 
.
ATTRIBUTE TYPEATTRIBUTE VALUE
_
Availabilitysystem/core-os
.TE

.SH SEE ALSO
.sp
.LP
\fBattributes\fR(5)
.SH NOTES
.sp
.LP
The default prompt for \fBcktime\fR is:
.sp
.in +2
.nf
Enter a time of day [?,q]:
.fi
.in -2
.sp

.sp
.LP
The default error message is:
.sp
.in +2
.nf
ERROR: Please enter the time of day.  Format is <format>.
.fi
.in -2
.sp

.sp
.LP
The default help message is:
.sp
.in +2
.nf
Please enter the time of day.  Format is <format>.
.fi
.in -2
.sp

.sp
.LP
When the quit option is chosen (and allowed), \fBq\fR is returned along with the return code \fB3\fR. The \fBvaltime\fR module will not produce any output. It returns  \fB0\fR for success and non-zero for failure.


usr/src/lib/libadm/common/cktime.c
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License (the "License").
 * You may not use this file except in compliance with the License.
 *
 * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*	Copyright (c) 1984, 1986, 1987, 1988, 1989 AT&T	*/
/*	  All Rights Reserved  	*/


/*
 * Copyright (c) 1989, 2011, Oracle and/or its affiliates. All rights reserved.
 */

#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <limits.h>
#include <sys/types.h>
#include <stdlib.h>
#include "libadm.h"

static int	fmtcheck(char *);

#define	PROMPT	"Enter the time of day"
#define	ERRMSG	"Please enter the time of day. Format is"
#define	DEFAULT	"%H:%M"

#define	TLEN 3
#define	LH 00
#define	UH 23
#define	USH 12
#define	LM 00
#define	UM 59
#define	LS 00
#define	US 59
#define	DELIM1 ':'
#define	BLANK ' '
#define	TAB '	'

static void
setmsg(char *msg, size_t msg_size, char *fmt)
{
	if (fmt == NULL)
		fmt = DEFAULT;
	(void) snprintf(msg, msg_size, "%s <%s>.", ERRMSG, fmt);
}

static char *
p_ndig(char *string, int *value)
{
	char *ptr;
	int accum = 0;
	int n = 2;

	if (!string)
		return (0);
	for (ptr = string; *ptr && n > 0; n--, ptr++) {
		if (! isdigit((unsigned char)*ptr))
			return (NULL);
		accum = (10 * accum) + (*ptr - '0');
		}
	if (n)
		return (NULL);
	*value = accum;
	return (ptr);
}

static char *
p_time(char *string, int llim, int ulim)
{
	char *ptr;
	int begin = -1;
	if (!(ptr = p_ndig(string, &begin)))
		return (NULL);
	if (begin >= llim && begin <= ulim)
		return (ptr);
	else return (NULL);
}

/* p_meridian will parse the string for the meridian - AM/PM or am/pm */

static char *
p_meridian(char *string)
{
	static char *middle[] = { "AM", "PM", "am", "pm" };
	int legit, n;
	char mid[TLEN];

	legit = 0;
	n = 0;
	mid[2] = '\0';
	(void) sscanf(string, "%2s", mid);
	while (!(legit) && (n < 4)) {
		if ((strncmp(mid, middle[n], 2)) == 0)
			legit = 1;	/* found legitimate string */
		n++;
	}
	if (legit)
		return (string+2);
	return (NULL);
}

static char *
p_delim(char *string, char dchoice)
{
	char dlm;

	if (! string)
		return (NULL);
	(void) sscanf(string, "%1c", &dlm);
	return ((dlm == dchoice) ? string + 1 : NULL);
}

int
cktime_val(char *fmt, char *input)
{
	char ltrl, dfl;
	int valid = 1; 	/* time of day string is valid for format */

	if ((fmt != NULL) && (fmtcheck(fmt) == 1))
		return (4);

	if (fmt == NULL)
		fmt = DEFAULT;
	ltrl = '\0';
	while (*fmt && valid) {
		if ((*fmt) == '%') {
			switch (*++fmt) {
			case 'H':
				input = p_time(input, LH, UH);
				if (!input)
					valid = 0;
				break;

			case 'M':
				input = p_time(input, LM, UM);
				if (!input)
					valid = 0;
				break;

			case 'S':
				input = p_time(input, LS, US);
				if (!input)
					valid = 0;
				break;

			case 'T':
				input = p_time(input, LH, UH);
				if (!input) {
					valid = 0;
					break;
				}

				input = p_delim(input, DELIM1);
				if (!input) {
					valid = 0;
					break;
				}
				input = p_time(input, LM, UM);
				if (!input) {
					valid = 0;
					break;
				}
				input = p_delim(input, DELIM1);
				if (!input) {
					valid = 0;
					break;
				}
				input = p_time(input, LS, US);
				if (!input)
					valid = 0;
				break;

			case 'R':
				input = p_time(input, LH, UH);
				if (!input) {
					valid = 0;
					break;
				}
				input = p_delim(input, DELIM1);
				if (!input) {
					valid = 0;
					break;
				}
				input = p_time(input, LM, UM);
				if (!input) {
					valid = 0;
					break;
				}
				break;

			case 'r':
				input = p_time(input, LH, USH);
				if (!input) {
					valid = 0;
					break;
				}
				input = p_delim(input, DELIM1);
				if (!input) {
					valid = 0;
					break;
				}
				input = p_time(input, LM, UM);
				if (!input) {
					valid = 0;
					break;
				}
				input = p_delim(input, DELIM1);
				if (!input) {
					valid = 0;
					break;
				}
				input = p_time(input, LS, US);
				if (!input) {
					valid = 0;
					break;
				}
				input = p_delim(input, BLANK);
				if (!input) {
					valid = 0;
					break;
				}
				input = p_meridian(input);
				if (!input)
					valid = 0;
				break;

			case 'I':
				input = p_time(input, LH, USH);
				if (!input)
					valid = 0;
				break;

			case 'p':
				input = p_meridian(input);
				if (!input)
					valid = 0;
				break;

			default:
				(void) sscanf(input++, "%1c", &ltrl);
			}
		} else {
			dfl = '\0';
			(void) sscanf(input, "%1c", &dfl);
			input++;
		}
		fmt++;
	}

	if (!(*fmt) && (input) && (*input))
		valid = 0;

	return ((valid == 0));
}

int
cktime_err(char *fmt, char *error)
{
	char	defmesg[128];

	if ((fmt != NULL) && (fmtcheck(fmt) == 1))
		return (4);
	setmsg(defmesg, sizeof (defmesg), fmt);
	puterror(stdout, defmesg, error);
	return (0);
}

int
cktime_hlp(char *fmt, char *help)
{
	char	defmesg[128];

	if ((fmt != NULL) && (fmtcheck(fmt) == 1))
		return (4);
	setmsg(defmesg, sizeof (defmesg), fmt);
	puthelp(stdout, defmesg, help);
	return (0);
}

/*
 *	A little state machine that checks out the format to
 *	make sure it is acceptable.
 *		return value 1: NG
 *		return value 0: OK
 */
int
fmtcheck(char *fmt)
{
	int	percent = 0;

	while (*fmt) {
		switch (*fmt++) {
			case '%': /* previous state must be start or letter */
				if (percent == 0)
					percent = 1;
				else
					return (1);
				break;
			case 'H': /* previous state must be "%" */
			case 'M':
			case 'S':
			case 'T':
			case 'R':
			case 'r':
			case 'I':
			case 'p':
				if (percent == 1)
					percent = 0;
				else
					return (1);
				break;
			case TAB: /* previous state must be start or letter */
			case BLANK:
			case DELIM1:
				if (percent == 1)
					return (1);
				break;
			default:
				return (1);
		}
	}
	return (percent);
}

int
cktime(char *tod, char *fmt, char *defstr, char *error, char *help,
	char *prompt)
{
	char	input[MAX_INPUT];
	char	defmesg[128];

	if ((fmt != NULL) && (fmtcheck(fmt) == 1))
		return (4);

	if (fmt == NULL)
		fmt = DEFAULT;
	setmsg(defmesg, sizeof (defmesg), fmt);
	if (!prompt)
		prompt = "Enter a time of day";

start:
	putprmpt(stderr, prompt, NULL, defstr);
	if (getinput(input))
		return (1);

	if (!strlen(input)) {
		if (defstr) {
			(void) strcpy(tod, defstr);
			return (0);
		}
		puterror(stderr, defmesg, error);
		goto start;
	}
	if (strcmp(input, "?") == 0) {
		puthelp(stderr, defmesg, help);
		goto start;
	}
	if (ckquit && (strcmp(input, "q") == 0))
		return (3);

	if (cktime_val(fmt, input)) {
		puterror(stderr, defmesg, error);
		goto start;
	}
	(void) strcpy(tod, input);
	return (0);
}
